### PR TITLE
Add RFD 183: Access Request Kube Resource Allow List

### PR DESCRIPTION
**Pull request by kimlisa**
[rendered](https://github.com/gravitational/teleport/blob/b64659fb1126f1d24fafbf8131576a207391efc4/rfd/0183-access-request-kube-resource-allow-list.md)

part of: https://github.com/gravitational/teleport/issues/46742

### CODE REVIEW

File: rfd/0183-access-request-kube-resource-allow-list.md [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+```yaml
+kind: role
+metadata:
+  name: requester
+spec:
+  allow:
+    request:
+      # Only allows requesting for kube_cluster's namespace resources
+      kubernetes:
+        allow: ['namespace']
```
Comment (r0mant) [inline rfd/0183-access-request-kube-resource-allow-list.md:114]: Putting this in the requester role still feels a bit off tbh, for a couple of reasons:

- The configuration for what you can actually request is now split between your requester role and the role you're requesting.
- This makes it harder to support scenarios like allowing requesters to request specific namespaces for one cluster, but everything (or another set of resources) for another cluster; you'd need to define multiple requester roles for that.

@kiosion is actually working on a somewhat similar feature request - making access request reason mandatory (https://github.com/gravitational/teleport/issues/20164), and we were considering making it an option on the _role being requested_ rather than requester role. Then you can have different requirements for resources granted by different roles.

Have you considered something similar here? Feels these two things are related and we should have a consistent approach. E.g. just off the top of my head if e.g. you have a role like this:

```yaml
spec:
  allow:
    kubernetes_resources:
      - kind: namespace
        name: pumpkin-*
        required: true
```

Then when someone is requesting the resource granted by this role, they have to select a namespace.

File: rfd/0183-access-request-kube-resource-allow-list.md [hunk 2]
```diff
@@ -0,0 +1,225 @@
+---
+author: Lisa Kim (lisa@goteleport.com)
+state: draft
+---
+
+# RFD 183 - Access Request Kubernetes Resource Allow List
+
+# Required Approvers
+
+- Engineering: @r0mant (roman) && @tigrato (tiago)
```
Comment (r0mant) [inline rfd/0183-access-request-kube-resource-allow-list.md:10]:
```suggestion
- Engineering: @r0mant (roman) && @tigrato (tiago)
- Product: @klizhentas || @xinding33
```

File: rfd/0183-access-request-kube-resource-allow-list.md [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+  allow:
+    request:
+      # Only allows requesting for kube_cluster's namespace resources
+      kubernetes:
+        allow: ['namespace']
+```
+
+## As an admin, I want to require users to specify a namespace from select few namespaces
+
+Admin's can combine the use of request kuberenetes allow list and `kubernetes_resources` field to customize what users can request to:
```
Comment (r0mant) [inline rfd/0183-access-request-kube-resource-allow-list.md:119]:
```suggestion
Admin's can combine the use of request Kubernetes allow list and `kubernetes_resources` field to customize what users can request to:
```

File: rfd/0183-access-request-kube-resource-allow-list.md [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+        name: pumpkin-*
+        verbs:
+          - list
+```
+
+# Details
+
+## New request field for role spec
+
+We will introduce a new role option under the `request` field `kuberenetes.allow` that holds a list of supported [Kubernetes resource kinds](https://github.com/gravitational/teleport/blob/110b23aefb3c4b91c9d8cca594041b93f0e078dd/api/types/constants.go#L1233).
```
Comment (r0mant) [inline rfd/0183-access-request-kube-resource-allow-list.md:143]:
```suggestion
We will introduce a new role option under the `request` field `kubernetes.allow` that holds a list of supported [Kubernetes resource kinds](https://github.com/gravitational/teleport/blob/110b23aefb3c4b91c9d8cca594041b93f0e078dd/api/types/constants.go#L1233).
```

**Comment by r0mant** [comment]
@klizhentas @xinding33 Can you review this when you get a chance?

File: rfd/0183-access-request-kube-resource-allow-list.md (hunk 1 above)
Comment (tigrato) [reply rfd/0183-access-request-kube-resource-allow-list.md:114]: My initial thought is that, since this feature is tied to Access Requests, we should clearly organize it under an "access request" block.

From my experience, customers often use roles as base roles for sysadmins or highly privileged users. This means that while some users might request access to role X, other users may be permanent members of that role.

With that in mind, the following configuration could be confusing:

```yaml
    kubernetes_resources:
      - kind: namespace
        name: pumpkin-*
        required: true
```

What does it mean for a permanent member of the role to have `required: true` in this context? I think this particular setting could be difficult to interpret.

I don’t dislike the idea of moving towards options for target roles, similar to this:

```yaml
kind: role
metadata:
  name: kube-access-request
spec:
  allow:
    request:
      search_as_roles:
      - devs
---
kind: role
version: v7
metadata:
  name: devs
spec:
  allow:
    logins: ['{{internal.logins}}']
    kubernetes_groups: ['{{internal.kubernetes_groups}}']
    node_labels:
      '*': '*'
    kubernetes_labels:
      '*': '*'
    kubernetes_resources:
      - kind: '*'
        name: '*'
        namespace: '*'
        verbs: ['*']
  options:
    access_request:
        reason: 'mandatory|optional'
        sub_resource: 'mandatory|optional'  # sub_resource is more generic and applies to namespaces, pods, secrets...
```

However, we need to be clear on the expected behavior when a user can request access with multiple roles that have conflicting options. Some of our options are "greedy" (where the most restrictive wins), while others only apply if the role is being used. I think we should define the expected behavior and user experience clearly in this case.


---

Placing it under the requestor role simplifies the behavior, as it provides a clear view of which roles can be requested and the rules governing the request, without needing to switch between different roles to assess the rules. For TAG, it’s also more straightforward to define the rules in the same location where we configure the `request` settings, similar to:

```yaml
kind: role
metadata:
  name: kube-access-request
spec:
  allow:
    request:
      options:
        reason: 'mandatory|optional'
        sub_resource: 'mandatory|optional'  # sub_resource is more generic and applies to namespaces, pods, secrets...
      search_as_roles:
      - devs
---
kind: role
version: v7
metadata:
  name: devs
spec:
  allow:
    logins: ['{{internal.logins}}']
    kubernetes_groups: ['{{internal.kubernetes_groups}}']
    node_labels:
      '*': '*'
    kubernetes_labels:
      '*': '*'
    kubernetes_resources:
      - kind: '*'
        name: '*'
        namespace: '*'
        verbs: ['*']
```

I am ok with any of the options above.

Comment (kimlisa) [reply rfd/0183-access-request-kube-resource-allow-list.md:114]: going back to roman's suggestion, can we do something like this:

```
>> This means user can request either kube_cluster or its subresources   (default)
 kubernetes_resources:
    - kind: '*'                   ←—
      name: '*'
      namespace: '*'
      verbs:
      - '*'

>> This means user must specify a namespace:
kubernetes_resources:
      - kind: 'namespace'   ←– 
        name: '*'
        verbs: ['*']

```

so basically, if we find that `kind != *`, then we will disable requesting for the whole cluster?

Comment (tigrato) [reply rfd/0183-access-request-kube-resource-allow-list.md:114]: Not exactly, because they may have a kind other than '*' and still need access to the entire cluster. For example, someone might need access to nodes, and certificate signing requests (which are global resources). You wouldn't be able to replicate this access pattern since we won’t initially support all of those resource types

Comment (kimlisa) [reply rfd/0183-access-request-kube-resource-allow-list.md:114]: i see.

i also think the main issue is the ability to `opt-in` to this feature (which won't disrupt current flows, and a admin might want to limit namespaces but allow whole cluster requests anyways). that means we _have_ to add a new field somewhere right?

if we go with roman's suggestion:

```
    kubernetes_resources:
      - kind: namespace
        name: pumpkin-*
        access_request_require: true     <--- opt-in
```

but if we are going to add a new field anyways, why not add a field in a central place? (whereever we end up putting it)

b/c i don't think it makes sense to do something like this:

```yaml
    kubernetes_resources:
     # restrict secrets
      - kind: secret
        access_request_require: true     # <--- disables whole cluster request
     # restrict pods
      - kind: pod
        access_request_require: false / true    # <--- isn't this redundant? or what does it mean if this means true as well?
```

Comment (r0mant) [reply rfd/0183-access-request-kube-resource-allow-list.md:114]: @kimlisa @tigrato To duplicate what we were discussing in Slack.

I was looking through our access requests documentation and configuration today and found some things that I didn't even know (or forgot) we support. Specifically, there are some additional request options that can be set on the requester role, for example: https://goteleport.com/docs/admin-guides/access-controls/access-requests/access-request-configuration/#how-clients-request-access. You can actually already specify things like "require request reason". It doesn't really work the way we need it to work for the thing @kiosion is working on (current impl. is tied to automatic access requests) but the point is that we already have something called [RequestAccess strategy](https://github.com/gravitational/teleport/blob/282d6d7ff5b1b8fc26da3d1c809685bcb75ea2fd/api/proto/teleport/legacy/types/types.proto#L2887-L2890), and a bunch of other options tied to the requester's role.

So with that in mind: I still find it more limiting than setting options on the roles being requested tbh but in order to not break any existing concepts and keep things consistent, I think I'm going to agree with the approach you both proposed initially and keeping this option as a part of the requester role. The only thing I think we still need to iron out is how to make the configuration more robust than just having a single `force_subresource_request` flag (which is a little confusing tbh) or even having `allow: ['namespace']` because that still doesn't allow any templating. So maybe we could do something like this:

```yaml
spec:
  allow:
    request:
      roles:
        - "dba"
      kubernetes:
        namespaces:
        - kube-system
        - {{external.kube_namespaces}}
        - *
```

Basically, if the namespaces option is present among the user's allow.request setting, we will force them to select namespaces when submitting an access request. If you, say, have a `*` there, you will be able to select any namespaces allowed by the role granting this cluster (assuming it also has `kube_resources` on it). Otherwise, you're limited to picking namespaces from the list of allowed ones in your request settings, which can also be templated from IdP attributes, like most other fields.

WDYGT?

Comment (kimlisa) [reply rfd/0183-access-request-kube-resource-allow-list.md]: summarizing our slack discussion here:

we should keep the ability to define limits to kube subresources with the `kubernetes_resources` field since it now also supports trait interpolation (coming in v17 release).

agreed on adding a role request option using enum values `cluster | resource | both`

Comment (kimlisa) [reply rfd/0183-access-request-kube-resource-allow-list.md]: updated RFD: https://github.com/gravitational/teleport/pull/46691/commits/26d187ff8611456d6b21b3ba9f28a71261d60df8

File: rfd/0183-access-request-kube-resource-allow-list.md [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+```yaml
+kind: role
+metadata:
+  name: role-name
+spec:
+  allow:
+    request:
+      kubernetes:
+        request_mode: cluster | resource | both
```
Comment (r0mant) [inline rfd/0183-access-request-kube-resource-allow-list.md:153]: I would convert it to a list with two modes indicating allowed request modes: `cluster` and `namespace`. Then you would be able to specify `request_modes: ["cluster", "namespace"]` (default) to allow both or individual ones which would be a slightly cleaner interface IMO.

Comment (kimlisa) [reply rfd/0183-access-request-kube-resource-allow-list.md]: i changed it to allow any sub resource kind: https://github.com/gravitational/teleport/blob/ebbdca1c30509f292deadce218d4b8f096a850b4/rfd/0183-access-request-kube-resource-allow-list.md#new-request-field-for-role-spec

File: rfd/0183-access-request-kube-resource-allow-list.md [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+
+| Request Mode Values                           |                           Explanation                            |
+| --------------------------------------------- | :--------------------------------------------------------------: |
+| []                                            |                         no restrictions                          |
+| `options.request_mode.kubernetes` not defined |                         no restrictions                          |
+| [kube_cluster, resource]                      |                         no restrictions                          |
+| [resource]                                    | requires users to request for any of the Kubernetes subresources |
+| [kube_cluster]                                |    requires users to request for the whole Kubernetes cluster    |
+| [namespace]                                   |          requires users to request for only namespaces           |
+| [namespace, pod]                              |      requires users to request for only namespaces OR pods       |
```
Comment (r0mant) [inline rfd/0183-access-request-kube-resource-allow-list.md:186]: The general idea makes sense but I feel like this is too many values tbh, I would scope this down to two possible options for now: `kube_cluster` and `namespace`. Then from user's perspective it will be clear what they can either allow someone to request access to cluster or namespace or both. We can always add more options later if needed but let's start simple.

Comment (kimlisa) [reply rfd/0183-access-request-kube-resource-allow-list.md]: replaced with sasha's recommendation

File: rfd/0183-access-request-kube-resource-allow-list.md [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+metadata:
+  name: requester
+spec:
+  options:
+    request_mode:
+      # Can request only pods or namespaces
+      kubernetes_resources:
+        - kind: namespace
+        - kind: pod
+```
```
Comment (nklaassen) [inline rfd/0183-access-request-kube-resource-allow-list.md:178]: I'm curious why we have this under `spec.options.request_mode.kubernetes_resources` instead of, say, `spec.allow.request.kubernetes_resources`. Basically why is it a role option and not under `spec.allow.request` where most other access request configuration is? And what is the effect if the option is set differently in multiple roles the user has?

I haven't really thought about it hard enough to say one of those options is definitively better, I'm just wondering if there's a reason or it just seemed like the default way to do it

Comment (nklaassen) [reply rfd/0183-access-request-kube-resource-allow-list.md:178]: After slightly more thought, I think this doesn't really make sense as a role option, and an allow/deny rule would be better

Here's my mental model of role options (I'm sure there are exceptions to this but i think it's mostly accurate). Usually, role options are coalesced across all roles a user has to result in a single value, and usually that value ends up encoded in the user's certificate. For example, the `forward_agent` option is combined from all roles the user has, and it's encoded in the SSH cert when the user logs in. The same for many other options. Even when not encoded in the cert, usually you compute a single value from the user's entire role set.

If instead you want one role to allow the user to request certain kubernetes namespaces with certain target roles, and another role to allow the user to request kubernetes clusters with a different target role, and another role can deny the user requesting a specific namespace, I think this would be better as an allow/deny rule under `spec.allow.request`

Comment (kimlisa) [reply rfd/0183-access-request-kube-resource-allow-list.md]: @nklaassen updated RFD to reflect field change

**Review by r0mant — APPROVED** [review]
@kimlisa lgtm with one syntactic comment

**Review by tigrato — APPROVED** [review]
